src/driver/mod.rs

```diff
@@ -43,30 +43,29 @@ pub struct Driver<E: Engine> {
     /// L1 chain watcher
     chain_watcher: ChainWatcher,
     /// Channel to receive the shutdown signal from
-    shutdown_recv: Receiver<bool>,
+    shutdown_recv: Arc<Receiver<bool>>,
 }
 
 impl Driver<EngineApi> {
-    pub async fn from_config(config: Config, shutdown_recv: Receiver<bool>) -> Result<Self> {
+    pub async fn from_config(config: Config, shutdown_recv: Arc<Receiver<bool>>) -> Result<Self> {
         let provider = Provider::try_from(&config.l2_rpc_url)?;
 
-        let head: Option<HeadInfo> = provider
-            .get_block_with_txs(BlockId::Number(BlockNumber::Finalized))
-            .await
-            .ok()
-            .flatten()
-            .and_then(|block| block.try_into().ok());
-
-        let finalized_head = head
-            .as_ref()
-            .map(|h| h.l2_block_info)
-            .unwrap_or(config.chain.l2_genesis);
+        let block_id = BlockId::Number(BlockNumber::Finalized);
+        let head = match HeadInfo::from_block(block_id, &provider).await? {
+            Some(head) => head,
```

Comment:
I have one doubt on this line: should we start from the execution layer `finalized` head after syncing?
I've noticed that if we do this after a full checkpoint sync cycle, it starts from the genesis block, wasting time.

Reply:
Yes, that's why I was surprised when it started doing this. I think it may have failed fetching that block and it fell back to that. I'm investigating

Reply:
I couldn't reproduce it, right now it's picking up the right block